docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,153 @@
+
+The watcher index is used to find watchers similarly. For example, on receiving an event containing object of labels “app:redis” and “role:slave”, storage doesn’t have to scan over all watchers, but only get those watching on these labels.
+
+In order to make use of indexes, we need select predicate to tell us three things: index key (a field or a label), operator (greater, less, or equal), and value (string, int, etc.). The original design of label and field has the three elements, e.g. “name=nginx”, “spec.nodename=A”. Although we had the correct design on the user side, somehow we lost it on server side in storage layer. The current storage layer can only get a filter or matching func that hides too many details and removes the ability to use indexing. Recent effort like MatcherIndex is not an indexing solution either, and making the matching and filter logic more convoluted. To fix this, we are proposing to have the labels and fields parsed as storage selectors -- a new design that provides the three elements, and pass them down to storage layer.
+
+Indexing is useless if we cannot handle generic use cases. One of the design goals is to support generic indexing. To achieve this, we are introducing a mechanism to extract field values from generic k8s objects. Because each k8s resource has its own type of schema, we are introducing a flexible object parsing function: FieldValueGetFunc. Each resource registry defines its own FieldValueGetFunc to parse the field of an object and return corresponding value.
+
+Now that we have the three basic elements to use the indexing, and also the method to build generic indexing, it’s easy to know how to index k8s objects. To look further, the above changes also enabled us to index watchers, enabling us to find out related watchers for each event quickly by using index.
+
+#### 2. Remove unnecessary helper functions from the interface itself
+
+We remove GuaranteedUpdate() from the Interface. It only uses the existing APIs and does not need to be part of the core storage API. One of the goals is to make storage API minimum and isolated to be more maintainable. So we decide to move it out of the Storage API and implement the same functionality as helper function. 
+
+``` go
+GuaranteedUpdate(Storage, ...)
+```
+
+We have split List into List and Wait. The previous List function will block until the result is as fresh as given version. This can be split into two functions, where one function does only one thing. The previous List behavior could be implemented in the following way:
+
+```go
+func WaitAndList(storage, version, ...) ... {
+	storage.Wait(version)
+	return storage.List(...)
+}
+```
+
+#### 3. Simplify conditional check
+
+We replace the UID filed in the precondition with version. UID and other fields all depends on the version of the objects. By comparing the versions between objects, we effectively compare its UID. We might add more preconditions if needed.
```

Comment:
Most fields are unique by version. We want to make core API simple and abstracted away unnecessary fields.

> where I want to ensure I am deleting object X (with uid 'abcdef'), but I don't care what version of it I am deleting.

UID and many other fields can be deducted from version, and we can provide some helper functions to simplify top level user's work.